perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AttributesManagerBlImpl.java

```diff
@@ -4436,82 +4428,119 @@ public void doTheMagic(PerunSession sess, Member member, boolean trueMagic) thro
+	 *
+	 * @return attribute after merging his value
+	 *
+	 * @throws InternalErrorException if one of mandatory objects is null or some internal problem has occured
+	 * @throws WrongAttributeValueException attribute value of set attribute is not correct
+	 * @throws WrongReferenceAttributeValueException any reference attribute value is not correct
+	 * @throws WrongAttributeAssignmentException if attribute is not from the same namespace defined by primaryHolder
+	 */
+	private Attribute mergeAttributeValue(PerunSession sess, Attribute attribute, PerunBean primaryHolder) throws InternalErrorException, WrongAttributeValueException,
+			WrongReferenceAttributeValueException, WrongAttributeAssignmentException {
+		//If attribute is null, throw an exception
+		if(attribute == null) throw new InternalErrorException("Can't merge null attribute with anything!");
+		if(primaryHolder == null) throw new InternalErrorException("Can't merge attribute value without notNull primaryHolder!");
+
+		//Get stored attribute in Perun
+		Attribute storedAttribute = null;
+		try {
+			if(primaryHolder instanceof User) {
+				storedAttribute = getPerunBl().getAttributesManagerBl().getAttribute(sess, (User) primaryHolder, attribute.getName());
+			} else if(primaryHolder instanceof Member) {
+				storedAttribute = getPerunBl().getAttributesManagerBl().getAttribute(sess, (Member) primaryHolder, attribute.getName());
+			} else {
+				throw new InternalErrorException("Primary holder for attribute is not supported: " + primaryHolder);
+			}
+		} catch (AttributeNotExistsException e) {
+			throw new ConsistencyErrorException(e);
+		}
+
+		//if attribute to merge has null value or it's value is same as stored attribute's value, return the stored attribute
+		convertEmptyAttrValueToNull(attribute);
+		if(attribute.getValue() == null || Objects.equals(attribute.getValue(), storedAttribute.getValue())) return storedAttribute;
+
+		// Check type ArrayList
+		if (attribute.getType().equals(ArrayList.class.getName()) || attribute.getType().equals(BeansUtils.largeArrayListClassName)) {
+			ArrayList<String> updatedList = (ArrayList<String>) storedAttribute.getValue();
+			// If there were someting then find values which haven't been already stored
+			if (updatedList != null) {
+				for (String value : ((ArrayList<String>) attribute.getValue())) {
+					if (!updatedList.contains(value)) {
+						updatedList.add(value);
+					}
+				}
+				attribute.setValue(updatedList);
+			}
+		// Check type LinkedHashMap
+		} else if (attribute.getType().equals(LinkedHashMap.class.getName())) {
+			//Find values which haven't been already stored
+			LinkedHashMap<String, String> updatedMap = (LinkedHashMap<String, String>) storedAttribute.getValue();
+			if (updatedMap != null) {
+				LinkedHashMap<String, String> receivedMap = (LinkedHashMap<String, String>) attribute.getValue();
+				updatedMap.putAll(receivedMap);
```

Comment:
This means that if old map (updatedMap) and new map (recievedMap) have some same keys, values for such keys will be overwritten with corresponding value from new map.
This behavior is correct, but it's not intuitive. Therefore I would recommend to explicitly mention this in method's javadoc.

Reply:
Fixed in javadoc of this method.